onnxruntime/test/ir/graph_viewer_test.cc

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <iostream>
+#include "core/graph/graph_viewer.h"
+#include "test/providers/provider_test_utils.h"
+#include "test/util/include/asserts.h"
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+using namespace ONNX_NAMESPACE;
+namespace onnxruntime {
+namespace test {
+
+TEST(GraphViewer, FilteredGraph) {
+  auto model_uri = ORT_TSTR("testdata/scan_1.onnx");
+
+  std::shared_ptr<Model> p_model;
+  ASSERT_STATUS_OK(Model::Load(model_uri, p_model, nullptr, DefaultLoggingManager().DefaultLogger()));
+  Graph& graph = p_model->MainGraph();
+
+  // create a GraphViewer that filters to the first 2 nodes.
+  IndexedSubGraph subgraph;
+  std::unique_ptr<IndexedSubGraph::MetaDef> metadef = onnxruntime::make_unique<IndexedSubGraph::MetaDef>();
+  metadef->name = "TestSubgraph";
+  metadef->status = ONNX_NAMESPACE::EXPERIMENTAL;
+  metadef->since_version = 1;
+  std::unordered_set<std::string> outputs;
+  std::unordered_set<const NodeArg*> initializers;
+
+  auto add_inputs = [&](ConstPointerContainer<std::vector<NodeArg*>> defs) {
+    for (const auto* def : defs) {
+      if (def->Exists()) {
+        // not the output of a previous node
+        if (outputs.count(def->Name()) == 0) {
+          metadef->inputs.push_back(def->Name());
+        } else {
+          // consumed by node so no longer subgraph output
+          // NOTE: Ignoring edge case where a node output is an overall graph output AND a node input
+          outputs.erase(def->Name());
+        }
+
+        if (graph.IsInitializedTensor(def->Name())) {
+          initializers.insert(def);
+        }
+      }
+    }
+  };
+
+  auto add_node = [&](const Node& node) {
+    subgraph.nodes.push_back(node.Index());
+    add_inputs(node.InputDefs());
+    add_inputs(node.ImplicitInputDefs());
+
+    for (const auto* def : node.OutputDefs()) {
+      outputs.insert(def->Name());
+    }
+  };
+
+  // first node
+  auto nodes = graph.Nodes().cbegin();
+  add_node(*nodes++);
+
+  // second node
+  add_node(*nodes++);
+
+  subgraph.SetMetaDef(std::move(metadef));
+
+  GraphViewer viewer{graph, subgraph};
+
+  EXPECT_EQ(viewer.NumberOfNodes(), 2);
+  // we added nodes in the topo order so that should still be the case
+  EXPECT_THAT(subgraph.nodes, testing::ContainerEq(viewer.GetNodesInTopologicalOrder()));
+
+  nodes = graph.Nodes().cbegin();
+  auto end_nodes = graph.Nodes().cend();
```

Comment:
node and end_node?

Reply:
Renamed `nodes` to `cur_node`. Left as `end_nodes` as it's the iterator pointing to the end of the nodes marker and not the last node.